### PR TITLE
Reduce concept count by making all entity recognizers to be simple Recognizer

### PR DESCRIPTION
**Pull request by tomlm**
## Description
EntityRecognizers are simple wrappers around TextAnalytics APIs, but currently are only usable from within RegExRecognizer via the EntityRecognizerSet.  
This delta makes all EntityRecognizers full blown recognizers, allowing them to be composed like any other recognizer in the system. 

This allows composition like this to feed external recgonizer results into LUIS recognizer:
```javascript
{
    "$schema": "../schemas/app.schema",
    "$kind": "Microsoft.LuisRecognizer",
    "id": "LUIS_msmeeting-actions",
 ...
    "externalEntityRecognizer": {
        "$kind": "Microsoft.RecognizerSet",
        "recognizers": [
            { $kind": "Microsoft.PartNumberRecognizer" },
            { $kind": "Iciclecreek.QuotedTextEntityRecognizer" },
..
        ]
    }
}
```

## Specific Changes
* Changed EntityRecognizer base class 
  * to derive from Recognizer, so it can be used as a recognizer. 
  * implement standard RecgonizeAsync() method to return RecognizerResult data structure.
* Added **ClientMentionEntityRecognizer** to map client passed mention entities => RecognizerResult entity structure, allowing that data to be passed to LuisRecognizer.

## Testing
* entity recognizer tests which use the RecognizeAsync() method and RecognizerResult data structure

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizerSet.cs [hunk 1]
```diff
@@ -21,6 +22,10 @@ public class EntityRecognizerSet : List<EntityRecognizer>
         /// <summary>
         /// Initializes a new instance of the <see cref="EntityRecognizerSet"/> class.
         /// </summary>
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EmailEntityRecognizer"/> class.
```
Comment (EricDahlvang) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizerSet.cs:26]: Extra and incorrect summary

File: tests/tests.schema [hunk 2]
```diff
@@ -689,7 +692,10 @@
       }
     },
     "Microsoft.AgeEntityRecognizer": {
-      "$role": "implements(Microsoft.IEntityRecognizer)",
+      "$role": [
```
Comment (EricDahlvang) [inline tests/tests.schema:695]: I'm still coming up to speed on some of this schema functionality.  Are these schmea changes just additive and backwards compatible?  If this is merged, and I open a project created with 4.10 in 4.11, everything new just shows up without issue?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs [hunk 3]
```diff
@@ -1,27 +1,122 @@
...
 using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Bot.Builder.Dialogs.Adaptive.Recognizers
 {
     /// <summary>
     /// Entity recognizers base class.
     /// </summary>
-    public class EntityRecognizer
+    public class EntityRecognizer : Recognizer
```
Comment (EricDahlvang) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs:18]: Isn't this a non-backward compatible breaking change?

File: tests/tests.schema (hunk 2 above)
Comment (tomlm) [reply tests/tests.schema:695]: yes additive

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs (hunk 3 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs:18]: no, it's not breaking change.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizerSet.cs (hunk 1 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizerSet.cs:26]: good catch, fixed.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs [hunk 4]
```diff
@@ -1,27 +1,122 @@
...
+            : base(callerPath, callerLine)
         {
         }
 
+        /// <summary>
+        /// Runs current DialogContext.TurnContext.Activity through a recognizer and returns a <see cref="RecognizerResult"/>.
+        /// </summary>
+        /// <param name="dialogContext">The <see cref="DialogContext"/> for the current turn of conversation.</param>
+        /// <param name="activity"><see cref="Activity"/> to recognize.</param>
+        /// <param name="cancellationToken">Optional, <see cref="CancellationToken"/> of the task.</param>
```
Comment (EricDahlvang) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs:36]: cancellationToken is not optional on this method.  Should it be?

**Review by EricDahlvang — APPROVED** [review]

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
+﻿using System.Collections.Generic;
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs:1]: Copyright header

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+using Microsoft.Bot.Schema;
+using Microsoft.Recognizers.Text;
+using Microsoft.Recognizers.Text.Sequence;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive.Recognizers
+{
+    /// <summary>
+    /// Recognizer which maps entities passed from the client activity.Entities (@type=mention) into RecognizerResult format.
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs:15]: if possible add linked docs

```suggestion
    /// Recognizer which maps entities passed from the client activity.Entities (@type=mention) into <see cref="RecognizerResult" /> format.
```

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            if (activity.Entities != null)
+            {
+                if (result.Entities == null)
+                {
+                    result.Entities = new JObject();
+                }
+
+                dynamic entities = result.Entities;
+
+                // convert activity.entities entity that looks like this:
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs:54]: Thanks for sharing the format :)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                    //            "text": "@tom",
+                    //            "type": "mention",
+                    //            "resolution": {
+                    //              "value": "@tom"
+                    //            }
+                    //         }
+                    //      ]
+                    //   }
+                    // }
+                    if (entities.mention == null)
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ClientMentionEntityRecognizer.cs:86]: I wish we didn't have to do so much `JToken` manipulation but I am not sure it can be avoided specifically for this scenario.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs (hunk 4 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs:36]: And if it's not optional, we should fix the comment.

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/EntityRecognizerRecognizerTests.cs [hunk 9]
```diff
@@ -0,0 +1,372 @@
+﻿using System;
```
Comment (carlosscastro) [inline tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/EntityRecognizerRecognizerTests.cs:1]: Copyright header

**Review by carlosscastro — APPROVED** [review]

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ChannelMentionEntityRecognizer.cs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+                    //     "mention": [
+                    //        {
+                    //            "startIndex": 10,
+                    //            "endIndex": 13,
+                    //            "score": 1.0,
+                    //            "text": "@tom",
+                    //            "type": "mention",
+                    //            "resolution": {
+                    //              "value": "@tom"
+                    //            }
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/ChannelMentionEntityRecognizer.cs:82]: Why is this here?  If it is really a resolution belongs in entities.mention which should be a structure and otherwise isn't it just the text?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizerSet.cs [hunk 11]
```diff
@@ -1,6 +1,10 @@
-﻿using System;
+﻿// Licensed under the MIT License.
+// Copyright (c) Microsoft Corporation. All rights reserved.
+
+using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Threading.Tasks;
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizerSet.cs:8]: You aren't using this because you don't have the debugger information.  Either drop it or use it.

**Comment by chrimc62** [comment]
>         public TextEntity()

Do you need caller information?

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/TextEntity.cs:22 in 6bea634. [](commit_id = 6bea63414e87481af5d347b7b45f44bb9f0abdd1, deletion_comment = False)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/EntityRecognizers/Microsoft.ChannelMentionEntityRecognizer.schema [hunk 12]
```diff
@@ -0,0 +1,8 @@
+{
+    "$schema": "https://schemas.botframework.com/schemas/component/v1.0/component.schema",
+    "$role": [ "implements(Microsoft.IRecognizer)" ],
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/EntityRecognizers/Microsoft.ChannelMentionEntityRecognizer.schema:3]: >implements [](start = 16, length = 10)

Do you want to implement IEntityRecognizer as well?  Or are you deprecating this?

File: tests/tests.schema [hunk 13]
```diff
@@ -3864,6 +3923,63 @@
         },
         {
           "$ref": "#/definitions/Microsoft.RegexRecognizer"
+        },
+        {
+          "$ref": "#/definitions/Microsoft.AgeEntityRecognizer"
+        },
+        {
```
Comment (chrimc62) [inline tests/tests.schema:3930]: Hmm.  These are supposed to be alphabetical.  This might be a bug in dialog:merge.

**Review by chrimc62 — CHANGES_REQUESTED** [review]
:clock1:

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/EntityRecognizers/Microsoft.ChannelMentionEntityRecognizer.schema (hunk 12 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Recognizers/EntityRecognizers/Microsoft.ChannelMentionEntityRecognizer.schema:3]: this isn't an entityrecognizer, it's only recgonizing data from the activity structure, not from the .text (or other entities itself).

---
In reply to: [511175574](https://github.com/microsoft/botbuilder-dotnet/pull/4848#discussion_r511175574) [](ancestors = 511175574)

**Comment by tomlm** [comment]
>         public TextEntity()

no 

---
In reply to: [715610450](https://github.com/microsoft/botbuilder-dotnet/pull/4848#issuecomment-715610450) [](ancestors = 715610450)

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/TextEntity.cs:22 in 6bea634. [](commit_id = 6bea63414e87481af5d347b7b45f44bb9f0abdd1, deletion_comment = False)

**Review by chrimc62 — APPROVED** [review]
:shipit:
